templates/support/OAuthAccessTokenManager.swift.hbs

```diff
@@ -0,0 +1,126 @@
+//  
+//  {{>frag/generatedBy}}
+//
+
+/** A token manager that will group refresh attempts to prevent concurrent token requests */
+import Foundation
+
+actor OAuthAccessTokenManager {
+    
+    var accessToken: OAuthToken? {
+        didSet {
+            restartRefreshTimer()
+        }
+    }
+    
+    private let refreshTokenURL: URL?
+    private let automaticallyRefreshTokensThisManySecondsEarly: TimeInterval?
+    private var refreshTask: Task<(), Error>?
+    private var refreshTimer: Timer?
+    private let prepareRequest: ((_ request: URLRequest) -> URLRequest)?
+    
+    init(refreshTokenURL: URL?, automaticallyRefreshTokensThisManySecondsEarly: TimeInterval? = nil, prepareRequest: ((_: URLRequest) -> URLRequest)? = nil) {
+        self.automaticallyRefreshTokensThisManySecondsEarly = automaticallyRefreshTokensThisManySecondsEarly
+        self.prepareRequest = prepareRequest
+        self.refreshTokenURL = refreshTokenURL
+    }
+        
+    private func restartRefreshTimer() {
+        refreshTimer?.invalidate()
+        refreshTimer = nil
+        
+        guard 
+            let autoTime = automaticallyRefreshTokensThisManySecondsEarly,
+            let token = accessToken, 
+            let expiresIn = token.expiresIn,
+            expiresIn - autoTime > 0
+        else {
+            return
+        }
+        
+        let timeout = Date(timeIntervalSinceNow: expiresIn - autoTime)
+        refreshTimer = Timer.scheduledTimer(withTimeInterval: expiresIn - autoTime, repeats: false, block: { t in
+            Task {
+                do {
+                    try await self.refreshToken()
+                } catch {
+                    print("An error ocurred in the automatic refresh: \(error)")
```

Comment:
This feels like it's only for debugging... our assumption is that the app will then use an API and get the refresh failure reported in a thrown error that they handle eh? So I guess it's OK to have it here? Maybe we can use the new OSLog thing so it looks a little less like it was just left behind?

Reply:
The logger needed would look something like this 
```swift
@available(macOS 11.0, iOS 14.0, watchOS 7.0, tvOS 14.0, *)
private let logger = Logger(subsystem: "{{@root.generatorClass}}", category: "oauth")
```
do you have a better subsystem name?

Reply:
Perhaps `@root.info.title`

Reply:
`@root.info.title` is empty, perhaps it's not being ingested properly from the yaml? Furthermore, apple suggests a reverse domain style. Perhaps we could have a name like `swift-client-generator.{{@root.info.title}}` might even want to convert that title to a classname version of it?